flink-download-jar-runtime/src/main/java/basic/application/StreamingJob.java

```diff
@@ -0,0 +1,262 @@
+
+		DataStream<String> inputStream = tokenizedStream
+				.keyBy(event -> event.f0)
+				.window(TumblingEventTimeWindows.of(Time.seconds(15)))
+				.sum(1)
+				.map(value -> value.f0 + " Count: " + value.f1 + "\n");
+
+		inputStream.addSink(createS3SinkFromStaticConfig(s3SinkPath));
+		env.execute("Flink Streaming Java API Skeleton");
+	}
+
+	private static void downloadUserJars(String region, String userJarKey,String userJarFileURI, String userJarBucket) throws IOException {
+		AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
+				.withRegion(region)
+				.withCredentials(new DefaultAWSCredentialsProviderChain())
+				.build();
+		// Get an object and print its contents.
+		LOG.info("Downloading " + userJarKey);
+
+		S3Object fileObject = s3Client.getObject(new GetObjectRequest(userJarBucket, userJarKey));
+		InputStream in = new BufferedInputStream(fileObject.getObjectContent());
+
+		ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+		byte[] buf = new byte[1024];
+
+		int n = 0;
+
+		while (-1 != (n = in.read(buf))) {
+
+			out.write(buf, 0, n);
+
+		}
+
+		out.close();
+		in.close();
+		byte[] response = out.toByteArray();
+		FileOutputStream fos = new FileOutputStream(userJarFileURI);
+		fos.write(response);
+		fos.close();
+	}
+
+	public static final class CustomTokenizer extends RichFlatMapFunction<String, String> {
+
+		@Override
+		public void open(Configuration parameters) throws Exception {
+			super.open(parameters);
+		}
+
+		@Override
+		public void flatMap(String s, Collector<String> collector) throws Exception {
+
+		}
+	}
+
+	// you don't need this process function if the JAR you're downloading only needs to be referenced in the main method
+	// ... and not within the operators.
+	public static final class ProcessTokenizer  extends ProcessFunction<String, Tuple2<String, Integer>> {
+		private final String userJarFileName;
+
+		public ProcessTokenizer(String userJarFileName) {
+			this.userJarFileName = userJarFileName;
+		}
+		/** The state that is maintained by this process function */
+		private URLClassLoader userCodeClassLoader;
+
+		// This method will be called once each time the job starts or restarts.
+		//... you can use the URLClassLoader directly in main method if you don't need to reference the
+		//... downloaded JAR in the stream processing operators
+		@Override
+		public void open(Configuration parameters) throws Exception {
+
+			File jar = getRuntimeContext().getDistributedCache().getFile(userJarFileName);
+			ClassLoader original = getRuntimeContext().getUserCodeClassLoader();
+			userCodeClassLoader = new URLClassLoader(new URL[]{ jar.toURL() }, original);
+		}
+
+		@Override
+		public void processElement(String s, ProcessFunction<String, Tuple2<String, Integer>>.Context context, Collector<Tuple2<String, Integer>> collector) throws Exception {
+
+			// Get the reference to the class
+			Class clazz = userCodeClassLoader.loadClass("com.myapp.core.utils.EventData");
```

Comment:
Why is this class and method instantiated on the hot path, ie, for every element that is processed. Would it be possible to avoid this redundant effort by moving the initialization into the `open` method?

Reply:
Moved to the open method